core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -3039,6 +3040,113 @@ protected final void createAggImpl(
     }
   }
 
+  /**
+   * The {@code GROUP_ID()} function is used to distinguish duplicate groups.
+   * However, as Aggregate normalizes group sets (i.e., sorting, redundancy removal),
+   * this information is lost in RelNode. Therefore, it is impossible to
+   * implement the function in runtime.
+   *
+   * To fill this gap, an aggregation query that contains {@code GROUP_ID()} function
+   * will generally be rewritten into UNION when converting to RelNode.
+   *
+   * Also see the discussion in JIRA
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1824">[CALCITE-1824]
+   * GROUP_ID returns wrong result</a>.
+   */
+  private RelNode rewriteAggregateWithGroupId(Blackboard bb,
+      AggregatingSelectScope.Resolved r, AggConverter converter) {
+    final List<AggregateCall> aggregateCalls = converter.getAggCalls();
+    final ImmutableBitSet groupSet = r.groupSet;
+    final Map<ImmutableBitSet, Integer> groupSetCount = r.groupSetCount;
+
+    final List<String> fieldNamesIfNoRewrite = createAggregate(bb, groupSet,
+        r.groupSets, aggregateCalls).getRowType().getFieldNames();
+
+    // For every GROUP_ID value, collect its group sets in map
+    // E.g., GROUPING SETS (a, a, b, c, c, c, c), the map will be
+    // {0 -> (a, b, c), 1 -> (a, c), 2 -> (c), 3 -> (c)},
```

Comment:
What does `(a, a, b, c, c, c, c)` mean here? How is it connected with `(a,b,c)` and `(a,c)`?

Reply:
They are duplicate group set before transformation into to canonical form `grouping sets (a, b, c)`. The `UnionAll`'s children and corresponding `group_id` is:
```
Union All
    grouping sets(a, b, c) --- 0
    grouping sets(a, c)     --- 1
    grouping sets(c)         --- 2
    grouping sets(c)         --- 3
```

Reply:
Note: the current comment says *what* the code is doing.
However, it would be better to clarify *why* it is doing like that.

The map itself has no value here. The value comes from splitting the grouping sets into the least number of `union all`. This intention is not really represented.

Reply:
Thanks a lot. I update the comment.